### ISSUE TITLE
metamorphic test failure: inconsistent keys in range

### ISSUE THREAD

**Issue by RaduBerinde** [cockroachdb/pebble#3174]
CI ran into a metamorphic test failure in PR #3173. It's unlikely (but not impossible) that the PR itself introduces the problem (it's mostly code cleanup).

https://github.com/cockroachdb/pebble/actions/runs/7266582327/job/19798611718?pr=3173

```

		1703013741613279788
		===== DIFF =====
		_meta/231219-192221.6133753307642/{standard-000,random-009}
		@@ -2103,11 +2103,11 @@
 iter20.SetBounds("rvxxunegblh@1", "tjpwuigvwge@1") // <nil> #2102
 iter20.SeekLT("tjpwuigvwge@1", "") // [true,"tjaj",<no point>,["tjaj","tjpwuigvwge@1")=>{"@6"="gqxiqxgpiskukyar"}*] <nil> #2103
 iter20.Next("") // [false] <nil> #2104
 iter20.Next("") // [false] <nil> #2105
 snap17.Close() // <nil> #2106
-iter22.Next("") // [true,"rpsllbp",<no point>,["rpsllbp","rvxxunegblh")=>{"@6"="iclhhyqvjdtfc","@4"="qumipltbbtehcoala","@3"="gtsviaw"}*] <nil> #2107
+iter22.Next("") // [true,"rpsllbp",<no point>,["rpsllbp","rvxxunegblh")=>{""="aevhlsxcjbxypnnas","@6"="iclhhyqvjdtfc","@4"="qumipltbbtehcoala","@3"="gtsviaw"}*] <nil> #2107
```

**Comment by RaduBerinde** [comment]
Reproduced on master (69994ddb7235eb42f5377f3af66556e57944ed07):

```
		===== SEED =====
		1703020073158492000
		===== DIFF =====
		_meta/231219-130753.1583218548355/{standard-000,random-027}
		@@ -4276,11 +4276,11 @@
 iter44.Prev("cksgzvdspi@8") // [valid,"gwmb",<no point>,["gwmb","jjbudo")=>{""="irymnotaxnqtzfxjdv","@9"="kyet","@6"=""}*] <nil> #4275
 batch30.Delete("zsgbcatts@6") // <nil> #4276
 db2.Ingest(batch30) // <nil> #4277
 iter44.SeekPrefixGE("zkqrrqt@5") // [false] pebble: SeekPrefixGE supplied with key outside of upper bound #4278
 iter44.Prev("") // [false] pebble: SeekPrefixGE supplied with key outside of upper bound #4279
-iter44.SeekPrefixGE("fusdchsqf@8") // [true,"fusdchsqf@8",<no point>,["fusdchsqf","fusdchsqf\x00")=>{"@9"="kyet","@6"=""}*] <nil> #4280
+iter44.SeekPrefixGE("fusdchsqf@8") // [true,"fusdchsqf@8",<no point>,["fusdchsqf","fusdchsqf\x00")=>{"@9"="kyet","@6"="","@5"="jmngn"}*] <nil> #4280
```

**Comment by nicktrav** [comment]
Can you share the repro command? I tried the following, but it doesn't repro for me on Darwin or Linux, on `master` @ 69994ddb7235eb42f5377f3af66556e57944ed07.

```console
$ CGO_ENABLED=0 go test -test.v -run TestMeta$ ./internal/metamorphic -seed 1703020073158492000 -keep
```

**Comment by itsbilal** [comment]
Looks like it's `TestMetaTwoInstance` that's affected, which is also the only metamorphic test failure to have VSSTs and related functionality enabled.

**Comment by nicktrav** [comment]
Ah, you're right. This does it:

```
$ go test -run TestMetaTwoInstance$ ./internal/metamorphic -seed 1703020073158492000
```

**Comment by RaduBerinde** [comment]
Interestingly I got a different error with that same seed:
```
                checker failed with error: out of order keys qtvsxobnl@1#294,DELSIZED >= qtvsxobnl@1#294,DELSIZED in L6: fileNum=000070
```

**Comment by itsbilal** [comment]
@RaduBerinde is this with your change to surface the start bound by any chance? I'm guessing not, but still good to know.

**Comment by RaduBerinde** [comment]
No, it is reproducible on master (see 2nd comment).

I realized we are not runnign `TestMetaTwoInstance` in the nightlies, filed #3175

**Comment by RaduBerinde** [comment]
I wrote a little tool to reduce the number of operations to get a small reproduction. I got one with these ops:
```
Init(2 /* dbs */, 50 /* batches */, 73 /* iters */, 50 /* snapshots */)
db2.DeleteRange("pdzxpawc@1", "zyltfzrkfx")
db2.Delete("qtvsxobnl@1")
snap3 = db2.NewSnapshot("cqlkmvj", "gwmb", "pdzxpawc", "pjtsroxxhzl", "rvutqvp", "tyklrgn", "vqdcha", "xmmrme", "ztgkqdcuwk", "zyltfzrkfx")
db2.Delete("qtvsxobnl@1")
db2.Compact("rvutqvp@1", "tyklrgn@1", false /* parallelize */)
db2.Replicate(db1, "gwmb", "rdiijw")
db1.Replicate(db2, "exiocx", "urziitjlj")
```

and these OPTIONS
```
[Version]
  pebble_version=0.1

[Options]
  bytes_per_sync=131072
  cache_size=64
  cleaner=delete
  compaction_debt_concurrency=1073741824
  comparer=pebble.internal.testkeys
  disable_wal=true
  flush_delay_delete_range=275ms
  flush_delay_range_key=685ms
  flush_split_bytes=16
  format_major_version=16
  l0_compaction_concurrency=1
  l0_compaction_file_threshold=1024
  l0_compaction_threshold=14
  l0_stop_writes_threshold=38
  lbase_max_bytes=4194304
  level_multiplier=160
  max_concurrent_compactions=3
  max_manifest_file_size=2
  max_open_files=1000
  mem_table_size=8192
  mem_table_stop_writes_threshold=2
  min_deletion_rate=1048576
  merger=pebble.concatenate
  multilevel_compaction_heuristic=wamp(0.00, false)
  read_compaction_rate=16000
  read_sampling_multiplier=16
  strict_wal_tail=true
  table_cache_shards=10
  table_property_collectors=[]
  validate_on_ingest=true
  wal_dir=data/wal
  wal_bytes_per_sync=0
  max_writer_concurrency=0
  force_writer_parallelism=false
  secondary_cache_size_bytes=33554432
  create_on_shared=1
  disable_elision_only_compactions=true

[Level "0"]
  block_restart_interval=58
  block_size=2
  block_size_threshold=89
  compression=NoCompression
  filter_policy=testing_bloom_filter/bits_per_key=16
  filter_type=table
  index_block_size=4
  target_file_size=1024

[TestOptions]
  delete_sized=true
  replace_single_delete=true
  threads=10
  async_apply_to_db=true
  shared_storage_enabled=true
  secondary_cache_enabled=true
  seed_efos=11947775591189849618
  use_shared_replicate=true
  efos_always_creates_iterators=true
```

Error is:
```
db1.Replicate(db2, "exiocx", "urziitjlj") // <nil> #7
Seed: 0
checker failed with error: out of order keys qtvsxobnl@1#13,DELSIZED >= qtvsxobnl@1#13,DELSIZED in L6: fileNum=000009
```

**Comment by itsbilal** [comment]
@RaduBerinde nice find! It looks like "boomerang files" which are files that get shared to another pebble, and then shared back to the same pebble, need to also have `hideObsoletePoints` set to true in `tableCacheShard.newIters`:

https://github.com/cockroachdb/pebble/blob/5be92739e7bb3cc2bb2158ad14be32bf7bc19e9f/table_cache.go#L552C1-L552C1

I think we'd want to set that to `true` for any shared ingested files (and we can determine ingested by seeing if the start/end seqnums are the same), not just shared foreign files. That also mimics the way we track ingested files for substituting sequence numbers, so is easier to reason about.

**Comment by RaduBerinde** [comment]
Thanks. Yes, that seems to fix the problem. I am however running in other problems related to table format versions being too low. I'm working on a change to require VirtualSSTables format to use shared objects.

**Comment by RaduBerinde** [comment]
I posted PR #3179 for the "key out of order" issue. I can still repro a compare failure. I will extend the "reduce" tool to work on compare as well.

**Comment by RaduBerinde** [comment]
I managed to reduce a repro to this:
```
Init(2 /* dbs */, 49 /* batches */, 63 /* iters */, 45 /* snapshots */)
db2.RangeKeySet("hesu", "olwqrm", "@6", "ybdmeuxousglnwebka")
snap9 = db2.NewSnapshot("boqrkyiv", "fhgoonxbav", "iifk", "jhbioo", "lipbcfybquf", "nvolewnkbok", "olwqrm", "vqzangor")
db2.RangeKeyDelete("iifk", "njojeaszbsh")
db2.Compact("nvolewnkbok@2", "uunagerm@9", true /* parallelize */)
db2.Replicate(db1, "faaopmh", "njojeaszbsh")
iter25 = db1.NewIter("", "", 2 /* key types */, 0, 0, false /* use L6 filters */, "@7" /* masking suffix */)
iter25.SeekGE("lipbcfybquf@3", "")
```

```diff
-iter25.SeekGE("lipbcfybquf@3", "") // [false] <nil> #7
+iter25.SeekGE("lipbcfybquf@3", "") // [true,"lipbcfybquf@3",<no point>,["hesu","njojeaszbsh")=>{"@6"="ybdmeuxousglnwebka"}*] <nil> #7
```

**Comment by RaduBerinde** [comment]
I simplified the keys; the repro is now:

```
Init(2 /* dbs */, 49 /* batches */, 63 /* iters */, 45 /* snapshots */)
db2.RangeKeySet("c", "h", "@6", "foo")
snap9 = db2.NewSnapshot("a", "z")
db2.RangeKeyDelete("d", "f")
db2.Compact("g@2", "i@9", true /* parallelize */)
db2.Replicate(db1, "b", "f")
iter25 = db1.NewIter("", "", 2 /* key types */, 0, 0, false /* use L6 filters */, "@7" /* masking suffix */)
iter25.SeekGE("e@3", "")
```

This is a visual of the operations:
```
               |-----------------------------| RangeKeySet
   |-----------------------------------------------| Snapshot
                     |-----------| RangeKeyDelete
                                       |-----------| Compact
         |-----------------------| Replicate
                           |-> SeekGE
   a --- b --- c --- d --- e --- f --- g --- h --- i
```

When running with these [OPTIONS](https://gist.github.com/RaduBerinde/f38b5e3d2eb502fb4c1da1340a8518c0), we get:

```
iter25.SeekGE("e@3", "") // [true,"e@3",<no point>,["c","f")=>{"@6"="foo"}*] <nil> #7
```

This is incorrect, as this portion of the RangeKeySet has been removed by the RangeKeyDel.


The Replicate op involves sharing and virtualizing an SST file. The physical SST file contains these range keys (note, this is after seqnum rewriting):
```
c-d:{(#10,RANGEKEYSET,@6,foo)}
d-f:{(#10,RANGEKEYDEL) (#10,RANGEKEYSET,@6,foo)}
f-h:{(#10,RANGEKEYSET,@6,foo)}
```
[ Question: the snapshot prevented the RANGEKEYSET from being removed when writing out the SST file, correct? ]

The virtual raw range key iter contains:
```
c-d:{(#10,RANGEKEYSET,@6,foo)}
d-f:{(#10,RANGEKEYDEL) (#10,RANGEKEYSET,@6,foo)}
```

Tracking what our hefty iterator stack is doing is pretty hard, but it looks like the `keyspan.MergingIter` is calling [UserIteratorConfig.Transform](https://github.com/cockroachdb/pebble/blob/master/internal/rangekey/coalesce.go#L184) which skips RANGEKEYDEL keys.


The snapshot is important for the reproduction. Without the snapshot, the SST contains nicely resolved fragments:
```
c-d:{(#10,RANGEKEYSET,@6,foo)}
d-f:{(#10,RANGEKEYDEL)}
f-h:{(#10,RANGEKEYSET,@6,foo)}
```

Also, if we query the original `db2`, we get the correct result `iter25.SeekGE("e@3", "") // [true,"f",<no point>,["f","h")=>{"@6"="foo"}*] <nil> #7`. In this case `MegingIter->Transform->colapse` does the right thing, because the `Delete` key shadows the other key (as per [this invariant](https://github.com/cockroachdb/pebble/blob/master/internal/rangekey/coalesce.go#L304)).

It feels that the problem stems from the fact that we rewrite the sequence numbers to the same value, and that we have keys that are supposed to shadow other keys. With the new sequence numbers, the RangeKeyDelete does not shadow the RangeKeySet because they have the same sequence number.

@itsbilal what do you think? Perhaps we never want to share SSTs that have unresolved keys because of snapshots?

Also, what assertions are we missing? We should somehow check that shared SSTs have no shadowed keys?

**Comment by itsbilal** [comment]
@RaduBerinde thank you for digging into this and for the clear and concise explanation of what's going on!

As we discussed on Slack, the best approach at addressing this is to use the hideObsolete logic for block iters to hide away the keys that were only preserved for snapshots. I made #3187 which applies that fix. In your example the shadowed RANGEKEYSET would have its obsolete bit set, and when that sstable gets shared, the receiver will hide away the RANGEKEYSET before applying the sequence number substitution.

I'll have to think more about what assertions we can add that'd harden this. It might be tricky but we can probably throw something into `UserIteratorConfig` that catches this case for shared ingested files only.
